torchgeo/datasets/splits.py

```diff
@@ -0,0 +1,375 @@
+
+    hits = [
+        hits[i]
+        for i in randperm(sum(lengths), generator=generator)  # type: ignore[arg-type]
+    ]
+
+    new_indexes = [
+        Index(interleaved=False, properties=Property(dimension=3)) for _ in lengths
+    ]
+
+    for i, length in enumerate(lengths):
+        for j in range(length):  # type: ignore[arg-type]
+            hit = hits.pop()
+            new_indexes[i].insert(j, hit.bounds, hit.object)
+
+    return [_create_geodataset_like(dataset, index) for index in new_indexes]
+
+
+def random_bbox_splitting(
+    dataset: GeoDataset,
+    fractions: Sequence[float],
+    generator: Optional[Generator] = default_generator,
+) -> List[GeoDataset]:
+    """Split a GeoDataset randomly splitting its index's BoundingBoxes.
+
+    This function will go through each BoundingBox in the GeoDataset's index and
+    split it in a random direction.
+
+    Args:
+        dataset: dataset to be split
+        fractions: fractions of splits to be produced
+        generator: (optional) generator used for the random permutation
+
+    Returns
+        A list of the subset datasets.
+
+    .. versionadded:: 0.4
+    """
+    if sum(fractions) != 1:
+        raise ValueError("Sum of input fractions must equal 1.")
+
+    if any(n <= 0 for n in fractions):
+        raise ValueError("All items in input fractions must be greater than 0.")
+
+    new_indexes = [
+        Index(interleaved=False, properties=Property(dimension=3)) for _ in fractions
+    ]
+
+    for i, hit in enumerate(
+        dataset.index.intersection(dataset.index.bounds, objects=True)
+    ):
+        box = BoundingBox(*hit.bounds)
+        fraction_left = 1.0
+
+        horizontal, flip = randint(0, 2, (2,), generator=generator)
+        for j, frac in enumerate(fractions):
+
+            if fraction_left == frac:
+                new_box = box
+            elif flip:
+                box, new_box = box.split(
+                    (fraction_left - frac) / fraction_left, horizontal
+                )
+            else:
+                new_box, box = box.split(frac / fraction_left, horizontal)
+
+            new_indexes[j].insert(i, tuple(new_box), hit.object)
+            fraction_left -= frac
+            horizontal = not horizontal
+
+    return [_create_geodataset_like(dataset, index) for index in new_indexes]
+
+
+def random_grid_cell_assignment(
+    dataset: GeoDataset,
+    fractions: Sequence[float],
+    grid_size: int = 6,
+    generator: Optional[Generator] = default_generator,
+) -> List[GeoDataset]:
+    """Overlays a grid over a GeoDataset and randomly assigns cells to new GeoDatasets.
+
+    This function will go through each BoundingBox in the GeoDataset's index, overlay
+    a grid over it, and randomly assign each cell to new GeoDatasets.
+
+    Args:
+        dataset: dataset to be split
+        fractions: fractions of splits to be produced
+        grid_size: (optional) size of the grid
```

Comment:
We should clarify whether size here means "number of grid cells per image" or "number of pixels per grid cell" or "size in meters of each grid cell". We had a similar confusion with our samplers when we first rolled them out. I'm guessing it's the first one, but we should clarify in the documentation which we mean. I don't _think_ we will need to support the other options, but if we do we can figure that out in a different PR.